[IMP] website_event: change the configure tickets button display

Mariska-Archielli opened the pull request:
Change the display of this button so it's integrated in the tickets menu.

Task ID: 2459617

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/66482).

awa-odoo commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -154,19 +156,25 @@
                         <i class="fa fa-ban mr-2"/>Sold Out
                     </span>
-                    <div t-if="event.is_participating" id="registered" class="ml-auto pr-3 d-none">
+                    <div t-if="event.is_participating" id="registered" class="ml-auto pr-3 d-none o_toggle-d-none">
```
Doesn't seem to work the way you want it to, stays displayed when expanded:
![image](https://user-images.githubusercontent.com/41282536/113983293-098a6a00-984a-11eb-8990-fa5f1c23ae3f.png)

Old version doesn't show it when expanded:
![image](https://user-images.githubusercontent.com/41282536/113983344-17d88600-984a-11eb-82c7-e3995b18dc67.png)

Or maybe that's something we want to show all the time now?
But then why the "o_toggle-d-none" thing?

EDIT: just noticed the cross to close is replaced by a "tickets" button on my first screenshot??
Seems to happen if I expand the button too quickly (old version seems to have that issue as well).
Really strange.

awa-odoo commented on addons/website_event/static/src/scss/event_templates_page.scss:
```diff
@@ -34,24 +34,23 @@
 
-    .o_wevent_registration_btn {
-        flex: 0 0 100%;
+    .o_wevent_registration_title {
```
I don't understand why we give it the "btn" class just to hardcode default cursor and no click behavior behind it.

awa-odoo commented on addons/website_event/static/src/js/website_event_ticket_details.js:
```diff
@@ -38,9 +38,9 @@ odoo.define('website_event.ticket_details', function (require) {
-                $(ev.currentTarget).siblings().toggleClass('d-none');
-                this.$('.close').toggleClass('d-none');
+                $(ev.currentTarget).toggleClass('btn-primary');
+                $(ev.currentTarget).siblings('.o_toggle-d-none').toggleClass('d-none');
```
I'm not a fan of that magic ".o_toggle-d-none" class.
It makes the code a bit cryptic.

At this point we are better of specifically targeting the things we want to show/hide by giving them a specific class.
(While you're at it, this method could be a bit documented to explain what it does too).

Mariska-Archielli replied:
The method now targets directly impacted elements.
Comments have been added on the JS.

awa-odoo reviewed (commented):
Hello @Mariska-Archielli 

That's quite the diff "just" to change the look of a button, but that's mainly due to the multiple layouts depending on tickets count / collapse behavior / ...
I added some questions / strange behaviors to check before we can move on.

Thanks!

Mariska-Archielli commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -116,21 +116,23 @@
+                        <span class="" t-esc="event.start_sale_datetime"
                             t-options="{'widget': 'datetime', 'tz_name': event.date_tz, 'format': 'short'}"/>
-                    <span t-esc="event.date_tz"/>
+                        <span t-esc="event.date_tz"/><t t-esc="event.start_sale_datetime"/>
```
Ok yes, that was an error in the rebase resolution

awa-odoo commented on addons/website_event/static/src/js/website_event_ticket_details.js:
```diff
@@ -35,12 +35,16 @@ odoo.define('website_event.ticket_details', function (require) {
          * @private
          * @param {*} ev
          */
+         /* When the "Fold Tickets Details" option is active, this will be called each
+         time the user expand or fold the tickets (o_wevent_registration_btn). This
+         allows to show/hide elements depending on the folding state. */
```
This comments actually goes within the existing comment section (right above the `* @private` line).

This is how comments are written in JavaScript and how we are able to extract them using automated tools.

awa-odoo commented on addons/website_event/static/src/js/website_event_ticket_details.js:
```diff
@@ -35,12 +35,16 @@ odoo.define('website_event.ticket_details', function (require) {
-                $(ev.currentTarget).toggleClass('btn-primary text-left pl-0');
-                $(ev.currentTarget).siblings().toggleClass('d-none');
-                this.$('.close').toggleClass('d-none');
+                $(ev.currentTarget).toggleClass('btn-primary');
```
Suggestion:

```js
let $target = $(ev.currentTarget);
$target.toggleClass('btn-primary');
$target.children().toggleClass('d-none');
$target.siblings('.o_wevent_registration_title, .o_price_range').toggleClass('d-none');
```

Creating a variable when you use the same call multiple (3+) times is usually easier for readability.
And since the "siblings" method takes a selector, using a coma separator should select everything you need in one line (please re-test to make sure it works though).

awa-odoo commented on addons/website_event/static/src/scss/event_templates_page.scss:
```diff
@@ -34,24 +34,21 @@
+    .o_wevent_registration_title {
+        font-weight: $font-weight-bold;
+        text-transform: uppercase;
+        font-family: "Source Sans Pro", "Odoo Unicode Support Noto", sans-serif !important;
```
We usually avoid hardcoding font-families as those come from the selected theme.

Mariska-Archielli replied:
Because I changed the type of this title (it was a button, now it is a text), I couldn't get to have the same typo as before. Do you have any other way for me to do so ?

awa-odoo replied:
Simply don't do it IMO.
If it's a text, then it should be styled as a text according to the selected template.
So I would remove this font-family rule.

awa-odoo commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -403,27 +407,37 @@
     <xpath expr="//div[@id='o_wevent_tickets_collapse']" position="attributes">
         <attribute name="class">collapse</attribute>
     </xpath>
+    <xpath expr="//span[@id='tickets-title']" position="attributes">
+        <attribute name="class">py-2 o_wevent_registration_title text-left d-none</attribute>
+    </xpath>
```
I know the old code had ID based selectors, but I'm not a fan of those, not nicely prefixed and a bit messy.

Maybe you could do that instead:
```xml
<xpath expr="//span[hasclass('o_wevent_registration_title')]" position="attributes">
    <attribute name="class" add="d-none" separator=" "/>
</xpath>
```

Mariska-Archielli replied:
I don't mind using 'hasclass('o_wevent_registration_title')' in this situation, but I'll still have to use the id for the price-range bellow ? (see in the other comment: can't just get rid of it)

awa-odoo commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -403,27 +407,37 @@
+    <xpath expr="//span[@id='tickets-title']" position="attributes">
+        <attribute name="class">py-2 o_wevent_registration_title text-left d-none</attribute>
+    </xpath>
+    <xpath expr="//div[@id='price-range']" position="attributes">
+        <attribute name="class">o_price_range</attribute>
+    </xpath>
```
I think you can safely add this class directly into the DOM item to simplify.
Even if it's actually only used when the option is ticked.

And you can rename to `o_wevent_price_range` to keep the same prefix as others.

Mariska-Archielli replied:
I'll rename. However, I'm not only adding this class in this case, I'm also replacing the "d-none". For the current logic, I can't just get rid of this xpath, so I figured I might as well just write the right classes where it belongs

awa-odoo replied:
Mmh then I would rather have "o_wevent_price_range d-none" on the base DOM element and make an override based on the class.
The "o_wevent_price_range" is used to target this element so I think the code is more clear if it's always there (even if only used when the folding option is enabled).

awa-odoo reviewed (commented):
Hello @Mariska-Archielli 

Looking better already.
Added a few more comments to polish the code.

For your commit message, I think you could elaborate a bit more.
Things like:
> "The 'Configure Tickets' button has been extracted to a template to ease re-usability".

And

> We also removed some seemingly useless steps during the end-user testings tours.

And always finished by linking the task ID

> Task-2459617

It can sound a bit like "useless blabla" but it's actually quite useful if someone browses the history and lands on your commit without having any idea what the task is about in the first place.
It's also usually a bit related to your diff size, if you only added a single class to the button (like "font-weight-bold"), making it a single line diff, a one-sentence commit message is plenty, but here you have to "justify" a bit all the changes.

Again, thanks for follow-up!

awa-odoo commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -154,8 +153,11 @@
-                    <div id="price-range" class="pr-3 d-none"/>
+                <div class="d-flex align-items-center py-2 pl-3 pr-2 border-bottom">
+                    <span id="tickets-title" class="py-2 o_wevent_registration_title text-left">Tickets</span>
+                    <div id="price-range" class="o_wevent_price_range d-none"/>
```
I think you can get rid if the id now.

awa-odoo commented on addons/website_event/views/event_templates_page_registration.xml:
```diff
@@ -154,8 +153,11 @@
-                <div class="d-flex justify-content-between align-items-center py-2 pl-3 pr-2 border-bottom">
-                    <div id="price-range" class="pr-3 d-none"/>
+                <div class="d-flex align-items-center py-2 pl-3 pr-2 border-bottom">
+                    <span id="tickets-title" class="py-2 o_wevent_registration_title text-left">Tickets</span>
```
Same here, if it's possible to get rid of the id it's better.

awa-odoo commented on addons/website_event_sale/views/website_event_templates.xml:
```diff
@@ -23,6 +23,9 @@
             <span t-esc="event.event_ticket_ids[-1].price_reduce" t-options="{'widget': 'monetary', 'display_currency': website.pricelist_id.currency_id}"/>
         </span>
     </xpath>
+    <xpath expr="//div[hasclass('o_wevent_price_range')]" position="attributes">
+        <attribute name="class" add="px-2" separator=" " />
+    </xpath>
```
Not sure I understand this "px-2" addition.

The only time it's applying is when the element is hidden:
`<div id="price-range" class="o_wevent_price_range d-none px-2">`

When you enable the "Fold ticket details", the override removes the "px-2" class:
`<div id="price-range" class="o_wevent_price_range">`

So it really looks like this is unused anyway?
Can you double-check and remove this if it's the case?

awa-odoo reviewed (commented):
Hello @Mariska-Archielli 

Here are a few more comments, we are getting close now.

About the "bug" you were not able to reproduce, I tried this morning and I still have the double "Registered" green tag on your branch:
![Selection_999(493)](https://user-images.githubusercontent.com/41282536/114665210-0802eb00-9cfd-11eb-911c-b6d7ffd6504d.png)

And it's definitely not happening in previous code version:
![Selection_999(492)](https://user-images.githubusercontent.com/41282536/114665054-d427c580-9cfc-11eb-8424-21a61d1a2a2d.png)

To reproduce I simply registered with the admin account to the event and it shows the double green badge.

Can you take a look?

Thanks for the follow-up!

awa-odoo commented:
@Mariska-Archielli 

Small side note, sentences in commit messages should go up to the 80 characters length mark.

So:
> The 'Configure Tickets' button has been extracted to a template to ease re-usability.
> The template's structure has been modified to allow improvement about this button's display.
> We also removed some seemingly useless steps during the end-suer testing tours.

Would become:

> The 'Configure Tickets' button has been extracted to a template to ease
> re-usability.
> The template's structure has been modified to allow improvement about this
> button's display.
> We also removed some seemingly useless steps during the end-user testing tours.

The goal is for it to be easy to read on github and in most text-based viewers ("tig", ...).

Feel free to update the message when you have some time.

Mariska-Archielli commented:
> @Mariska-Archielli
> 
> Small side note, sentences in commit messages should go up to the 80 characters length mark.
> 
> So:
> 
> > The 'Configure Tickets' button has been extracted to a template to ease re-usability.
> > The template's structure has been modified to allow improvement about this button's display.
> > We also removed some seemingly useless steps during the end-suer testing tours.
> 
> Would become:
> 
> > The 'Configure Tickets' button has been extracted to a template to ease
> > re-usability.
> > The template's structure has been modified to allow improvement about this
> > button's display.
> > We also removed some seemingly useless steps during the end-user testing tours.
> 
> The goal is for it to be easy to read on github and in most text-based viewers ("tig", ...).
> 
> Feel free to update the message when you have some time.

Ok done :)

tde-banana-odoo commented:
Why is it closed ?
